src/graph/unit_graph.cc

```diff
@@ -687,7 +736,9 @@ BoolArray UnitGraph::HasVertices(dgl_type_t vtype, IdArray vids) const {
 }
 
 bool UnitGraph::HasEdgeBetween(dgl_type_t etype, dgl_id_t src, dgl_id_t dst) const {
-  if (in_csr_) {
+  if (prefer_coo_) {
+    return GetCOO()->HasEdgeBetween(etype, src, dst);
+  } else if (in_csr_) {
```

Comment:
The `GetSuitableFormat` is actually quite complicated if we have a "lock format" semantic:
1. If the format is locked, return the locked format no matter what.
2. Otherwise, if the graph is hypersparse, return COO no matter what.
3. Otherwise, return the preferred format depending on the API itself (e.g. `HasEdgeBetween` would go for whatever the graph has, `InEdges` would prefer CSC, `FindEdges` would prefer COO).

Does that make sense?

Reply:
There are two decisions that should not be mixed.
1. Which format to store the graph?
2. Which format to use for certain operator?

A good place for the first decision is the constructor function of a graph object, while the second decision is performed in each operator. As a result, if a graph is hypersparse, the constructor will decide to use COO and lock the format. Thus, my `GetSuitableFormat` is simpler as shown in the demo code.

Reply:
> As a result, if a graph is hypersparse, the constructor will decide to use COO and lock the format.

Should we lock whenever the graph is hypersparse (and not allow converting to CSR)?  What if I want to do message passing?  I'm afraid that we would break the code that indeed does message passing on a hypersparse graph dataset.

Reply:
> > As a result, if a graph is hypersparse, the constructor will decide to use COO and lock the format.
> 
> Should we lock whenever the graph is hypersparse (and not allow converting to CSR)? What if I want to do message passing? I'm afraid that we would break the code that indeed does message passing on a hypersparse graph dataset.

I c. The case is that some operators do not have implementation on a certain format while we still want to keep the code runnable, where locking a format is indeed not good. Can we change the `store_format` as `preferred_format` in the constructor, and the first argument of `GetSuitableFormat` as `allowed_format`:
(1) The `allowed_format` is a list of formats. Default is `kAny`.
(2) If `allowed_format` is specified, try match one with the stored format. Otherwise, convert and return any of the `allowed_format`.

Several cases:
* `HasEdgesBetween` provides `kAny`.
* `InEdges` provides `kCOO` and `kCSC`.
* `BinaryReduce` provides `kCSR` and `kCSC`.

What do you think?

Reply:
Note that this proposal does not have the `lock_format_` member anymore. The constructor directly stores the graph in the `preferred_format`.

Reply:
One thing came to me is the design of preferred_format of lock_format may conflict with the target_parallel kernel design. But I think we can delay completing the target parallel kernel implementation.

Reply:
It shouldn't? Kernel should specify which format it requests right?

Reply:
> It shouldn't? Kernel should specify which format it requests right?
Yes, kernel will specify the required format. Usually it requires all COO CSR CSC in case of forward and backward. So kernel need temp store for these formats.
Currently, kernel directly using low-level APIs of Unigraph e.g., getCSRMatrix(). So these ops will not be affected?

Another concern is which kind of kernel is good for hypersparse graph, which may be critical for GNN on large graph with sampling.

Reply:
> Yes, kernel will specify the required format. Usually it requires all COO CSR CSC in case of forward and backward. So kernel need temp store for these formats.

My hunch is SPMM/SDDMM kernels probably should not require all formats. Correct me if my understanding is wrong.
* For SPMM-like by node parallelism, CSC is required.
* For SDDMM-like by edge parallelism, COO is required.
* For backward SPMM by node parallelism, CSR is required.

> Another concern is which kind of kernel is good for hypersparse graph, which may be critical for GNN on large graph with sampling.

Sampled hypersparse subgraphs will be compacted before message passing so normal SPMM and SDDMM kernels should work fine.

Reply:
> I c. The case is that some operators do not have implementation on a certain format while we still want to keep the code runnable, where locking a format is indeed not good. Can we change the `store_format` as `preferred_format` in the constructor, and the first argument of `GetSuitableFormat` as `allowed_format`:
> (1) The `allowed_format` is a list of formats. Default is `kAny`.
> (2) If `allowed_format` is specified, try match one with the stored format. Otherwise, convert and return any of the `allowed_format`.
> 
> Several cases:
> 
> * `HasEdgesBetween` provides `kAny`.
> * `InEdges` provides `kCOO` and `kCSC`.
> * `BinaryReduce` provides `kCSR` and `kCSC`.
> 
> What do you think?

I adopted this solution (kind of).  Essentially we now

* have a `restrict_format_` member that restricts the format to run *graph queries* on.
* have `restrict_format_` set to COO automatically if the graph is hypersparse and constructed via COO (if from CSR we probably don't care since the indptr is already initialized anyway).
* have a private method `UnitGraph::GetFormat` which takes in a format suggestion (e.g. CSC) and return a storage pointer either as restricted, or as suggested if not restricted.
* allow materializing CSR/CSC/COO even if it is restricted, if requested explicitly (i.e. calling `GetInCSR`/`GetOutCSR`/`GetCOO` directly, bypassing `GetFormat`).

Reply:
> > Yes, kernel will specify the required format. Usually it requires all COO CSR CSC in case of forward and backward. So kernel need temp store for these formats.
> 
> My hunch is SPMM/SDDMM kernels probably should not require all formats. Correct me if my understanding is wrong.
> 
> * For SPMM-like by node parallelism, CSC is required.
> * For SDDMM-like by edge parallelism, COO is required.
> * For backward SPMM by node parallelism, CSR is required.
> 

consider  u_add_v none, in forward, it is edge parallel, COO is required, in backward, both lhs and rhs are node parallel, and with different target, both CSC and CSR is required. We need to take both forward and backward as a round-trip.

> > Another concern is which kind of kernel is good for hypersparse graph, which may be critical for GNN on large graph with sampling.
> 
> Sampled hypersparse subgraphs will be compacted before message passing so normal SPMM and SDDMM kernels should work fine.

I see. Compact do help.

Reply:
> * have a private method `UnitGraph::GetFormat` which takes in a format suggestion (e.g. CSC) and return a storage pointer either as restricted, or as suggested if not restricted.  All graph queries would get storage using `GetFormat`.
> * allow materializing CSR/CSC/COO even if it is restricted, if requested explicitly (i.e. calling `GetInCSR`/`GetOutCSR`/`GetCOO` directly, bypassing `GetFormat`).  This case is for message passing.

Are they conflicting with each other? The first bullet should be another API than `UnitGraph::GetFormat` which I suppose return the requested format no matter what.

Reply:
I suppose we are on the same page; My `GetFormat` is essentially your "another API", which I can rename to `GetSuitableFormat(SparseFormat suggestion)`.  Your `GetFormat` would be my `GetInCSR`/`GetOutCSR`/`GetCOO`.

Reply:
Yes, they are named `SelectFormat` and `GetFormat` respectively.